In [1]:
import pandas as pd
import numpy as np
import polars as pl

from m4_feats_polars import *
from m5_sb_models import *
import warnings
warnings.filterwarnings("ignore")

In [2]:
# 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
data_path     = 'kaggle/input/linking-writing-processes-to-writing-quality/'
train_logs    = pl.scan_csv(f'{data_path}/train_logs.csv')
test_logs    = pl.scan_csv(f'{data_path}/test_logs.csv')
train_scores = pl.scan_csv(f'{data_path}/train_scores.csv')

# train_logs, test_logs = amend_event_id_order(train_logs, test_logs)     worsens

In [3]:
def word_pauses(train_logs, test_logs):
    print("word pauses")    
    feats = []

    for data in [train_logs, test_logs]:
        logs = data.clone()
        logs = logs.with_columns(pl.col('word_count')
            .diff().over('id')
            .alias('word_diff'))

        logs = logs.filter(
            pl.col('word_diff')>0).select(pl.col(['id','action_time']))

        word_pause = logs.group_by(['id']).agg(
                        word_pause_count = pl.col('action_time').count(),
                        word_pause_mean = pl.col('action_time').mean(),
                        word_pause_sum = pl.col('action_time').sum(),
                        word_pause_std = pl.col('action_time').std(),
                        word_pause_max = pl.col('action_time').max(),
                        word_pause_min = pl.col('action_time').min(),
                        word_pause_median = pl.col('action_time').median(),
                        word_pasuse_q1 = pl.col('action_time').quantile(0.25),
                        word_pasuse_q3 = pl.col('action_time').quantile(0.75),
                        word_pasuse_kurt = pl.col('action_time').kurtosis(),
                        word_pasuse_skew = pl.col('action_time').skew(),
        )
        feats.append(word_pause)
    return feats[0], feats[1]

In [3]:
logs = train_logs.clone()
logs,_ = normalise_up_down_times(logs, test_logs)
logs = logs.select(pl.col(['id','event_id','word_count','down_time','up_time','action_time']))
logs = logs.with_columns(pl.col('word_count')
            .diff().over('id')
            .fill_null(1)
            .alias('word_diff'))

negs = logs.filter(pl.col('word_diff')<0)
pstv = logs.filter(pl.col('word_diff')>0)

In [4]:
logs.head(25).collect()

shape: (25, 7)
┌──────────┬──────────┬────────────┬───────────┬─────────┬─────────────┬───────────┐
│ id       ┆ event_id ┆ word_count ┆ down_time ┆ up_time ┆ action_time ┆ word_diff │
│ ---      ┆ ---      ┆ ---        ┆ ---       ┆ ---     ┆ ---         ┆ ---       │
│ str      ┆ i64      ┆ i64        ┆ i64       ┆ i64     ┆ i64         ┆ i64       │
╞══════════╪══════════╪════════════╪═══════════╪═════════╪═════════════╪═══════════╡
│ 001519c8 ┆ 1        ┆ 0          ┆ 0         ┆ 4557    ┆ 31          ┆ 1         │
│ 001519c8 ┆ 2        ┆ 0          ┆ 32        ┆ 4930    ┆ 404         ┆ 0         │
│ 001519c8 ┆ 3        ┆ 0          ┆ 102045    ┆ 4526    ┆ 0           ┆ 0         │
│ 001519c8 ┆ 4        ┆ 1          ┆ 102160    ┆ 4617    ┆ 91          ┆ 1         │
│ …        ┆ …        ┆ …          ┆ …         ┆ …       ┆ …           ┆ …         │
│ 001519c8 ┆ 22       ┆ 3          ┆ 106322    ┆ 4667    ┆ 141         ┆ 0         │
│ 001519c8 ┆ 23       ┆ 4          ┆ 106508    ┆ 4632    ┆ 106         ┆ 1         │
│ 001519c8 ┆ 24       ┆ 4          ┆ 106648    ┆ 4613    ┆ 87          ┆ 0         │
│ 001519c8 ┆ 25       ┆ 4          ┆ 106784    ┆ 4630    ┆ 104         ┆ 0         │
└──────────┴──────────┴────────────┴───────────┴─────────┴─────────────┴───────────┘

In [38]:
train_logs.head(25).collect()

shape: (25, 11)
┌──────────┬──────────┬───────────┬─────────┬───┬───────────┬────────────┬────────────┬────────────┐
│ id       ┆ event_id ┆ down_time ┆ up_time ┆ … ┆ up_event  ┆ text_chang ┆ cursor_pos ┆ word_count │
│ ---      ┆ ---      ┆ ---       ┆ ---     ┆   ┆ ---       ┆ e          ┆ ition      ┆ ---        │
│ str      ┆ i64      ┆ i64       ┆ i64     ┆   ┆ str       ┆ ---        ┆ ---        ┆ i64        │
│          ┆          ┆           ┆         ┆   ┆           ┆ str        ┆ i64        ┆            │
╞══════════╪══════════╪═══════════╪═════════╪═══╪═══════════╪════════════╪════════════╪════════════╡
│ 001519c8 ┆ 1        ┆ 4526      ┆ 4557    ┆ … ┆ Leftclick ┆ NoChange   ┆ 0          ┆ 0          │
│ 001519c8 ┆ 2        ┆ 4558      ┆ 4962    ┆ … ┆ Leftclick ┆ NoChange   ┆ 0          ┆ 0          │
│ 001519c8 ┆ 3        ┆ 106571    ┆ 106571  ┆ … ┆ Shift     ┆ NoChange   ┆ 0          ┆ 0          │
│ 001519c8 ┆ 4        ┆ 106686    ┆ 106777  ┆ … ┆ q         ┆ q          ┆ 1          ┆ 1          │
│ …        ┆ …        ┆ …         ┆ …       ┆ … ┆ …         ┆ …          ┆ …          ┆ …          │
│ 001519c8 ┆ 22       ┆ 110848    ┆ 110989  ┆ … ┆ Space     ┆            ┆ 17         ┆ 3          │
│ 001519c8 ┆ 23       ┆ 111034    ┆ 111140  ┆ … ┆ q         ┆ q          ┆ 18         ┆ 4          │
│ 001519c8 ┆ 24       ┆ 111174    ┆ 111261  ┆ … ┆ q         ┆ q          ┆ 19         ┆ 4          │
│ 001519c8 ┆ 25       ┆ 111310    ┆ 111414  ┆ … ┆ q         ┆ q          ┆ 20         ┆ 4          │
└──────────┴──────────┴───────────┴─────────┴───┴───────────┴────────────┴────────────┴────────────┘

In [3]:
# everything is logged - DONE
# bursts = 2/3 of a second - input only - DONE
# inter word pauses
# between sentence pauses ?
# between paragraph pauses ?
# backspace pauses
# edit pauses

In [5]:
# Best Feature Set: ['train_essay_sentences.pkl', 'train_create_pauses.pkl', 
# 'train_vector_one_gram.pkl', 'train_essay_paragraphs.pkl', 
# 'train_categorical_nunique.pkl', 'train_word_pauses.pkl', 
# 'train_events_counts_rate_of_change.pkl', 'train_word_counts_rate_of_change.pkl', 
# 'train_r_burst_feats.pkl', 'train_vector_two_gram.pkl', 'train_events_counts_acceleration.pkl']


In [6]:
# best_feature_set_1 - PARTIAL
train_essays          = get_essay_df(train_logs.collect().to_pandas())
test_essays           = get_essay_df(test_logs.collect().to_pandas())

tr_down_events_counts, ts_down_events_counts = down_events_counts(train_logs, test_logs)
tr_vect_one, ts_vect_one = countvectorize_one_one(train_essays, test_essays)
tr_pauses, ts_pauses = create_pauses(train_logs, test_logs)
tr_cursor_pos_acc, ts_cursor_pos_acc = cursor_pos_acceleration(train_logs, test_logs)
tr_word_pause, ts_word_pause = word_pauses(train_logs, test_logs)
tr_word_count_acc, ts_word_count_acc = word_count_acceleration(train_logs, test_logs)
#tr_p_burst, ts_p_burst = p_burst_feats(train_logs, test_logs, 2)
tr_r_burst, ts_r_burst = r_burst_feats(train_logs, test_logs)
#tr_event_acc, ts_event_acc = events_counts_acceleration(train_logs, test_logs)
# tr_nunique, ts_nunique = categorical_nunique(train_logs, test_logs)
tr_vect_two, ts_vect_two = countvectorize_two_one(train_essays, test_essays)
# tr_time_by_act, ts_time_by_act = action_time_by_activity(train_logs, test_logs)
# tr_cursor_pos_roc, ts_cursor_pos_roc = cursor_pos_rate_of_change(train_logs, test_logs)
# 
# tr_act_count, ts_act_count = count_of_activities(train_logs, test_logs)
# tr_get_keys, ts_get_keys = get_keys_pressed_per_second(train_logs.collect().to_pandas(), 
#                                                        test_logs.collect().to_pandas())
# 
# tr_input_change, ts_input_change = input_text_change_feats(train_logs, test_logs)
# tr_wc_roc, ts_wc_roc =  word_counts_rate_of_change(train_logs, test_logs)

train_feats = tr_down_events_counts.join(tr_vect_one, on='id', how='left')
train_feats = train_feats.join(tr_pauses, on='id', how='left')
train_feats = train_feats.join(tr_cursor_pos_acc, on='id', how='left')
train_feats = train_feats.join(tr_word_pause, on='id', how='left')
train_feats = train_feats.join(tr_word_count_acc, on='id', how='left')
#train_feats = train_feats.join(tr_p_burst, on='id', how='left')
train_feats = train_feats.join(tr_r_burst, on='id', how='left')
train_feats = train_feats.join(tr_vect_two, on='id', how='left')
# train_feats = train_feats.join(tr_event_acc, on='id', how='left')
# train_feats = train_feats.join(tr_nunique, on='id', how='left')
# train_feats = train_feats.join(tr_wc_roc, on='id', how='left')
# train_feats = train_feats.join(tr_act_count, on='id', how='left')
# train_feats = train_feats.join(tr_cursor_pos_roc, on='id', how='left')

# train_feats = train_feats.join(tr_get_keys, on='id', how='left')
# train_feats = train_feats.join(tr_input_change, on='id', how='left')
# train_feats = train_feats.join(tr_time_by_act, on='id', how='left')

test_feats = ts_down_events_counts.join(ts_vect_one, on='id', how='left')
test_feats = test_feats.join(ts_pauses, on='id', how='left')
test_feats = test_feats.join(ts_cursor_pos_acc, on='id', how='left')
test_feats = test_feats.join(ts_word_pause, on='id', how='left')
test_feats = test_feats.join(ts_word_count_acc, on='id', how='left')
# test_feats = test_feats.join(ts_p_burst, on='id', how='left')
test_feats = test_feats.join(ts_r_burst, on='id', how='left')
test_feats = test_feats.join(ts_vect_two, on='id', how='left')
# test_feats = test_feats.join(tr_event_acc, on='id', how='left')
# test_feats = test_feats.join(ts_nunique, on='id', how='left')
# test_feats = test_feats.join(ts_wc_roc, on='id', how='left')
# test_feats = test_feats.join(ts_act_count, on='id', how='left')
# test_feats = test_feats.join(ts_cursor_pos_roc, on='id', how='left')


# test_feats = test_feats.join(ts_get_keys, on='id', how='left')
# test_feats = test_feats.join(ts_input_change, on='id', how='left')
# test_feats = test_feats.join(ts_time_by_act, on='id', how='left')


train_logs = train_logs.collect().to_pandas()
test_logs = test_logs.collect().to_pandas()
train_scores = train_scores.collect().to_pandas()
train_feats = train_feats.sort('id')
train_feats = train_feats.collect().to_pandas()
test_feats = test_feats.collect().to_pandas()

train_feats           = train_feats.merge(parag_feats(train_essays), on='id', how='left')
test_feats            = test_feats.merge(parag_feats(test_essays), on='id', how='left')

train_feats           = train_feats.merge(sent_feats(train_essays), on='id', how='left')
test_feats            = test_feats.merge(sent_feats(test_essays), on='id', how='left')

train_feats = train_feats.merge(train_scores, on='id', how='left')
print(f'train feats shape {train_feats.shape}')


< Events counts features >
< Count vectorize one-grams >
< Idle time features >
< cursor position acceleration >
word pauses
< word count acceleration >
< R-burst features >
< Count vectorize bi-grams >
< Essays paragraphs feats >
< Essays paragraphs feats >
< Essays sentences feats >
< Essays sentences feats >
train feats shape (2471, 144)


In [7]:
from m5_sb_models import lgb_pipeline
lgb_params_1 = {
    'boosting_type': 'gbdt', 
    'metric': 'rmse',
    'reg_alpha': 0.0031, 
    'reg_lambda': 0.001, 
    'colsample_bytree': 0.8,  
    'subsample_freq': 1,  
    'subsample': 0.75,  
    'learning_rate': 0.017, 
    'num_leaves': 19, 
    'min_child_samples': 46,
    'n_estimators': 350,
    'verbosity': -1
    }

param = {'n_estimators': 1024,
        'learning_rate': 0.005,
        'metric': 'rmse',
        'force_col_wise': True,
        'verbosity': 0,}

# train_feats = train_feats[['id', 'score'] + feat_select]
# test_feats = test_feats[['id'] + feat_select]

print(f'train feats shape {train_feats.shape}')


train feats shape (2471, 144)


In [8]:
shuffle_preds = []

for i in range(15):
    train_feats = train_feats.sample(frac=1).reset_index(drop=True)
    test_preds, oof_preds, rmse, model = lgb_pipeline(train_feats, test_feats, lgb_params_1)
    shuffle_preds.append(rmse)
    #test_preds, oof_preds, rmse, model = lgb_pipeline(train_feats, test_feats, param)

np.mean(shuffle_preds)

Final RMSE over 50: 0.603617. Std 0.8270
RMSE by fold 0.603563. Std 0.0083
Final RMSE over 50: 0.601545. Std 0.8255
RMSE by fold 0.601473. Std 0.0091
Final RMSE over 50: 0.602574. Std 0.8276
RMSE by fold 0.602372. Std 0.0151
Final RMSE over 50: 0.601950. Std 0.8275
RMSE by fold 0.601897. Std 0.0080
Final RMSE over 50: 0.604154. Std 0.8274
RMSE by fold 0.603977. Std 0.0149
Final RMSE over 50: 0.602163. Std 0.8266
RMSE by fold 0.602061. Std 0.0108
Final RMSE over 50: 0.603044. Std 0.8271
RMSE by fold 0.602868. Std 0.0148
Final RMSE over 50: 0.602455. Std 0.8275
RMSE by fold 0.602338. Std 0.0117
Final RMSE over 50: 0.602454. Std 0.8279
RMSE by fold 0.602405. Std 0.0079
Final RMSE over 50: 0.602837. Std 0.8266
RMSE by fold 0.602635. Std 0.0155
Final RMSE over 50: 0.602299. Std 0.8262
RMSE by fold 0.602235. Std 0.0083
Final RMSE over 50: 0.602076. Std 0.8286
RMSE by fold 0.601953. Std 0.0120
Final RMSE over 50: 0.602816. Std 0.8265
RMSE by fold 0.602714. Std 0.0110
Final RMSE over 50: 0.602

0.6026620396931698

In [9]:
#  0.6046250116470197 - baseline
#  0.6034392851713597 - word_pauses
#  0.6031532703206033 - word_count_acc
#  0.6030545803940613 - r_burst
#  0.60246384515886   - bigrams
# sentences

- M4 + cursor_pos_acc + word_count_acc + bigrams <br />
0.6044949502480581


- M4 + cursor_pos_acc + word_count_acc  <br />
0.6047963953877546 <br />


In [10]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
# oof_res['rmse'] = oof_res.apply(lambda x: np.sqrt((x['score']-x['preds'])**2))
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)
oof_res.groupby(['score'])['RMSE'].mean().reset_index().sort_values('RMSE', ascending=False)

score      RMSE
0     0.5  1.528224
1     1.0  1.283878
11    6.0  1.132529
2     1.5  0.879977
10    5.5  0.738216
3     2.0  0.573584
9     5.0  0.474352
4     2.5  0.461494
5     3.0  0.460465
6     3.5  0.402527
8     4.5  0.355785
7     4.0  0.349586

In [11]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)
oof_res.groupby(['score'])['RMSE'].mean().reset_index().sort_values('RMSE', ascending=False)

score      RMSE
0     0.5  1.528224
1     1.0  1.283878
11    6.0  1.132529
2     1.5  0.879977
10    5.5  0.738216
3     2.0  0.573584
9     5.0  0.474352
4     2.5  0.461494
5     3.0  0.460465
6     3.5  0.402527
8     4.5  0.355785
7     4.0  0.349586

In [12]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
# oof_res['rmse'] = oof_res.apply(lambda x: np.sqrt((x['score']-x['preds'])**2))
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)

In [13]:
import lightgbm as lgb

weights = [1.0 for _ in range(len(y))]  # Default weights
weights[some_specific_index] = 1.5  # Higher weight for a specific instance

train_data = lgb.Dataset(X_train, label=y_train, weight=weights)
# Proceed with setting up parameters and training the model


NameError: name 'y' is not defined

train feats shape (2471, 151)
Final RMSE over 50: 0.605768. Std 0.8244
RMSE by fold 0.605635. Std 0.0128

reduced feats
Final RMSE over 50: 0.604472. Std 0.8245
RMSE by fold 0.604365. Std 0.0115

r burst only
Number of estimators: 350
Final RMSE over 50: 0.604057. Std 0.8244
RMSE by fold 0.603946. Std 0.0116

Number of estimators: 350
Final RMSE over 50: 0.604650. Std 0.8243
RMSE by fold 0.604527. Std 0.0122

In [ ]:
test_ids = test_feats.id
y_pred = np.mean(test_preds, axis=0)

sub = pd.DataFrame({'id': test_ids, 'score': y_pred})
sub.to_csv('submission.csv', index=False)